Implement BuyFeeAsset trait

cowboy-bebug opened the pull request:
Added:
- `primitives/src/traits.rs`
  + added BuyFeeAsset trait (accessible via cennznet_primitives::traits::BuyFeeAsset)
  + this is where commonly used traits in CENNZnet should live
- `crml/cennzx-spot/src/impls.rs`
  + added BuyFeeAsset impl
  + related tests are added
- `crml/cennzx-spot/src/mock.rs`
  + added FeeAssetCurrency for testing BuyFeeAsset trait impl
- `crml/cennzx-spot/src/types.rs`
  + added FeeExchange struct
- `crml/transaction-payment/src/lib.rs`
  + BuyFeeAsset associated type added to transaction_payment::Trait
- `runtime/src/lib.rs`
  + BuyFeeAsset added as an empty tuple in runtime logic

Closes #87

cowboy-bebug committed:
remove unsafe code block

- use to_le_bytes() to convert u64 -> byte array

cowboy-bebug committed:
add BuyFeeAsset trait in primitives/traits.rs

- common traits used by CENNZnet should live in primitives/traits.rs

cowboy-bebug committed:
add a placeholder impl for BuyFeeAsset

cowboy-bebug committed:
add FeeExchange struct and its method

cowboy-bebug committed:
add FEE_ASSET_ID and FeeAssetCurrency type to mock

cowboy-bebug committed:
implement BuyFeeAsset and reinstate associated tests

cowboy-bebug committed:
remove println used for debugging

cowboy-bebug committed:
add BuyFeeAsset trait to transaction-payment and runtime

- pair programmed with Hoani
- ran cargo fmt

aliXsed commented:
I reviewed the change more on the surface without digging in the logic. LGTM but leave the approval to @holygits

cowboy-bebug committed:
ensure fee rate is being tested in buy_fee_asset

- multiplied test currencies by 10
- added assert_eq! to assert_exchange_balance_eq! macro
- added `core_asset_price` and `trade_asset_price` to show calculations
- used variables to clarify the above calculations
- pair programmed with Hoani

hoani reviewed (approved):
Completed deep dive on BuyFeeAsset with Eric. LGTM
